walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,103 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.List;
+import java.util.Set;
 
 public final class FindMeetingQuery {
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+
+  /**
+   * Finds available time ranges for all requested attendees to attend the
+   * supplied MeetingRequest, given known events
+   * 
+   * @param events  All known events that will be considered
+   * @param request The meeting request whose attendees must all be free
+   * @return A collection of time ranges where all proposed attendees in @request
+   *         are not attending any known events in @events
+   */
+  public Collection<TimeRange> query(final Collection<Event> events, final MeetingRequest request) {
+    Collection<TimeRange> queryRanges = new ArrayList<>();
+    queryRanges.add(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, TimeRange.END_OF_DAY, true));
+    Map<Event, Map<MeetingRequest, Boolean>> cache = new HashMap<>();
+    for (Event event : events) {
```

Comment:
Since you run 1 pass through the events, every event will be accessed for a single time, so I'm not sure if you actually get a cache hit from the outer map.

What happens if you do not include the hash? Would the algorithm still work (pass all tests)?

Reply:
I was thinking that, since hasOverlappingAttendees can get called for multiple TimeRanges as we keep slicing up the available times, it becomes more useful as there are more TimeRanges. I can confirm that it does get hit, but not a lot, so it may not be worth the memory.

Reply:
What is the runtime of this method with the cache?